doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,245 @@
+This can easily explode script sizes.
+Therefore, only a few simple functions like `(&&)` and `(||)` should be defined this way.
+
+### Monomorphizing Plutus Tx Standard Library
+
+Haskell's `Data.Foldable.any` function has the following type:
+
+```haskell
+any :: forall t. Foldable t => (a -> Bool) -> t a -> Bool
+```
+
+Because Haskell is a non-strict language, it is easy to define `any` in terms of `foldr`, which guarantees to short circuit:
+
+```haskell
+any p = foldr f False
+  where f a acc = p a || acc
+```
+
+However, this definition of `any` does not work for Plutus Tx.
+Recall the definition of `PlutusTx.foldr` for lists:
+
+```haskell
+foldr f z = go
+  where
+    go = \case
+      []     -> z
+      x : xs -> f x (go xs)
+```
+
+Because Plutus Tx is strict, `go xs` will be evaluated before evaluating `f`, which means `any` cannot short circuit. And since, like said before, Plutus Tx lacks a good way to define arbitrary non-strict functions, given this definition of `foldr`, there is no way to implement an `any` function with the above type, which guarantees to short-circuit for every `t` that has a `Foldable` instance.
+
+We therefore choose to do the following for the Plutus Tx standard library:
+- Provide the `Foldable` class and instances as they are, but make it clear that they cannot short-circuit.
+- For functions that should short-circuit, such as `any` and `all`, we provide monomorphic versions of these functions (e.g., for lists), which are to be implemented without using `foldr`.
+
+This is the same as what some other strict languages do in their stdlibs, such as Scala and OCaml.
+
+It is worth mentioning that it is possible to implement a short-circuiting _and_ polymorphic `any`, if `Foldable` is defined in a different way, in which there are explicit suspensions and resumptions, e.g.,
+
+```haskell
+type Delay = (->) ()
+
+delay :: a -> Delay a
+delay a = \_ -> a
+
+force :: Delay a -> a
+force a = a ()
+
+class Foldable_Delay t where
+  foldr_delay :: (a -> Delay b -> Delay b) -> Delay b -> t a -> Delay b
+
+instance Foldable_Delay [] where
+  {-# INLINABLE foldr_delay #-}
+  foldr_delay f z = go
+    where
+      go = \case
+        [] -> z
+        x : xs -> f x (delay (force (go xs)))
+
+{-# INLINABLE any #-}
+any :: Foldable_Delay t => (a -> Bool) -> t a -> Bool
+any p = force . foldr_delay f (\_ -> False)
+  where
+    f a acc =
+      if p a then delay True
+      else acc
+```
+
+However, we will not be doing this in the standard library, because even though this version of `any` does short-circuit, it is not as efficient as the monomorphic version due to the additional cost from the `delay` and `force` machinery.
+If this version is indeed preferred for some reason, it can be provided in a third party library.
+
+### Turning off GHC's pre-inliner
+
+As said before, one reason why the semantics of Plutus Tx is difficult to predict is because we run GHC's pre-inliner, which may perform unconditionally inlining.
```

Comment:
```suggestion
As said before, one reason why the semantics of Plutus Tx is difficult to predict is because we run GHC's pre-inliner, which may perform unconditional inlining.
```